### PR TITLE
Add psbt BIP174 test vectors tests

### LINKED ISSUE #892: Write an example psbt workflow
Write an example in the `examples` directory with the workflow from [BIP174](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#test-vectors). See the valid example that starts from the line in the above link.
```
The private keys in the tests below are derived from the following master private key:
```
- Use BIP32 wallet APIs to derive all priv/public keys
- Use the sighash APIs to compute the signature hashes
- Use all Psbt supported APIs
- Identify the API pain-points and suggest improvements.

### PR DESCRIPTION
**Pull request by DanGould**
resolves #892

*  Initial patch adds OP_0 alias for OP_PUSHBYTES_0 as [approved here](https://github.com/rust-bitcoin/rust-bitcoin/pull/935#discussion_r854539416)
* Second patch is the bulk of this work. It tests BIP 174 PSBT integration defined at [The BIP's Test Vectors Section](https://github.com/bitcoin/bips/blob/master/bip-0174.mediawiki#test-vectors) using #957 
* Third patch points users to rust-miniscript for a more useful finalizer

### CODE REVIEW

**Comment by tcharding** [comment]
>  what do you mean by "Clean up this error" on SignError?

I stole that error type, and a bunch of the code for signing PSBTs from the [bdk](https://github.com/bitcoindevkit/bdk/) project. By 'clean up' I meant remove all the variants that were not being used. Also, in general, all the error handling code is incomplete i.e., usage of `expect` throughout. I was planning on re-doing all that once everything worked.

> One suggestion is to alias PartiallySignedTransaction as Psbt in rust-bitcoin

That's a PR you can do if you would like to. Conceptually simple but likely to get some argument/discussion if you like that sort of thing :) Otherwise I'll do it.

Cheers, will review.

**Review by tcharding — COMMENTED** [review]
I'm going to push a commit to this branch @DanGould, want to take a look at it and squash it in if you agree with the changes?

**Comment by tcharding** [comment]
Nice work man! We have the seed passing issue to resolve still. Also, I'm starting to think that this should not be an example but rather should be an integration test. Then, the remaining TODOs are changes to `rust-bitcoin`. If you want to do them I'm happy to let you run with them :)

**Comment by tcharding** [comment]
Locally this builds fine but CI is including recent changes to the sighash types, no idea what is going on with that. Its your branch @DanGould so I won't mess with it but my blunt guess would be update your local `master` branch then rebase this and change `SigHash` -> `Sighash` and force push.

File: examples/psbt.rs
Comment (DanGould) [inline examples/psbt.rs:155]: Parsing the seed has me stumped too. When I picked this up `seed` was being parsed from base64. That, nor base58, nor base58check result in a matching `ext_priv`. Is this example used in any other project or are we gonna have to ask achow how he did this? I checked the NBitcoin implementation where they list the seed as `masterseed` in a json file of test data but it is never consumed.

**Comment by DanGould** [comment]
> I'm starting to think that this should not be an example but rather should be an integration test —@tcharding

concept ack to that and your psbt watch-only example as an example. I'll turn these TODOs into PRs.

File: examples/psbt.rs
Comment (tcharding) [reply examples/psbt.rs:155]: Yeah I tried base58 too :) I asked on bitcoin-dev mailing list.

File: examples/psbt.rs
Comment (tcharding) [inline examples/psbt.rs:80]: These `QUESTION` are unanswered questions to myself, perhaps change this to `FIXME` with the aim that this gets resolved during review. Or perhaps `FIXME (Do not merge)` or something?

File: examples/psbt.rs
Comment (tcharding) [inline examples/psbt.rs:381]: This can be removed now, PSBT serialization is not deterministic. See https://github.com/rust-bitcoin/rust-bitcoin/issues/937. I think we should be doing all the `assert_eq!` statements with PSBTs like this one not with hex like all the others do so we do not falsely imply that hex serialization is deterministic.

File: examples/psbt.rs
Comment (tcharding) [inline examples/psbt.rs:454]: Warning, this TODO is non-trivial. The same signing code appears in #940, I'm not exactly sure right at this second how to handle this code or where it should go?

File: examples/psbt.rs
Comment (tcharding) [inline examples/psbt.rs:509]: The two checks above were just hacked together while looking at the test vector PSBTs, this code is not general enough (as I'm finding in #940), before the signing code can be considered ready for `rust-bitcoin` this needs improving and testing with other valid PSBTs, that in itself is an argument for leaving the signing code here in this file for now?

File: examples/psbt.rs
Comment (tcharding) [inline examples/psbt.rs:672]: I would like to know _why_ this code is needed and why its not in `rust-bitcoin`, said another way, why do we need to drop the first two bytes from the script?

**Comment by tcharding** [comment]
For your CI fixes you can check out #940 or 
- Use `BTreeMap:new()` and then call insert instead of using `BTreeMap::from`
- Add an explicit scope around the immutable/mutable borrow issue (if you hit it)

File: examples/psbt.rs
Comment (DanGould) [reply examples/psbt.rs:454]: We have signing code here because we're building a test or example rather than a library that does arbitrary signing. You can find [code that serves this purpose this in bdk](https://github.com/bitcoindevkit/bdk/blob/master/src/wallet/signer.rs). That seems to be place for it to me.

File: examples/psbt.rs
Comment (DanGould) [reply examples/psbt.rs:509]: I think it makes sense that the generalized code lives in bdk. The question is would using those in examples be a circular dependency. I don't think it makes sense for bdk code to live in integration tests.

Comment (tcharding) [reply examples/psbt.rs:505]: Oh we definitely cannot depend on bdk. What I mean is that I'm pretty sure this code should live in the `psbt` module but I'm not sure its complete enough yet or if it is, as it currently exists in this PR and #940, general enough. The main stuff I'm concerned about is the BIP174 checks that I added.

File: examples/psbt.rs
Comment (DanGould) [reply examples/psbt.rs:670]: We drop the first two bytes of the scriptPubKey as`script` because the first is the version (0x00) the second length of the pubkeyhash  (0x20) and the value p2wpkh script code wants here is the witness program AKA the pubkeyhash in p2wpkh

src: this [libbitcoin doc](https://github.com/libbitcoin/libbitcoin-system/wiki/P2WPKH-Transactions#spending-a-p2wpkh-output) and [the bip 141 spec](https://github.com/bitcoin/bips/blob/master/bip-0141.mediawiki).

Perhaps it should live in `rust-bitcoin` because it's de/serialization of standard script

File: examples/psbt.rs
Comment (tcharding) [reply examples/psbt.rs:155]: Got an answer from Andrew Chow, its formatted as a WIF private key, I have no idea why (I asked him that :).  The following code works
```
    // Strings from BIP174 test vector.
    let extended_private_key = "tprv8ZgxMBicQKsPd9TeAdPADNnSyH9SSUUbTVeFszDE23Ki6TBB5nCefAdHkK8Fm3qMQR6sHwA56zqRmKmxnHk37JkiFzvncDqoKmPWubu7hDF";
    let seed = "cUkG8i1RFfWGWy5ziR11zJ5V4U4W3viSFCfyJmZnvQaUsd1xuF3T";

    let xpriv = ExtendedPrivKey::from_str(extended_private_key).unwrap();

    let sk = PrivateKey::from_wif(&seed).unwrap();
    let seeded = ExtendedPrivKey::new_master(NETWORK, &sk.inner.secret_bytes()).unwrap();
    assert_eq!(xpriv, seeded);
```

File: examples/psbt.rs
Comment (DanGould) [reply examples/psbt.rs:505]: Why does this signer need to be general if it only serves to test PSBT? Is there an expectation it is library ready as well?

**Comment by DanGould** [comment]
Why the heck aren't I seeing these errors when running `act` locally? I feel like a pleb pushing CI failing builds

**Comment by tcharding** [comment]
Don't beat yourself up about it, it took me months on this project to work out how to see the Rust 1.29 errors. I don't use `act`, in case it helps I can get these errors locally using
```
RUST_LOG=warn cargo-gen-lockfile-rust-1.29-bitcoin; RUST_LOG=warn cargo +1.29 run --example psbt
```
And the shell function
```
function cargo-gen-lockfile-rust-1.29-bitcoin {
    RUST_LOG=warn cargo clean
    rm Cargo.lock

    cargo +1.29 generate-lockfile --verbose && \
        cargo +1.29 update -p cc --precise "1.0.41" --verbose && \
        cargo +1.29 update --package "serde" --precise "1.0.98" --verbose && \
        cargo +1.29 update -p "serde_derive" --precise "1.0.98" --verbose

    export RUST_LOG=$saved_log
}
```

**Comment by tcharding** [comment]
The CI failure is the scope one I mention above.

**Comment by DanGould** [comment]
@tcharding would it make sense to add this script to contributing.md? I wonder why this doesn't show up with act on its own. I get that this will change when we upgrade rust. this would have saved me 3 pushes at this point

**Comment by tcharding** [comment]
Yeah go for it.

File: examples/psbt.rs
Comment (tcharding) [reply examples/psbt.rs:502]: I was just wondering that's all, we are about to try to introduce duplicate code in this one and #940 which feels a bit rude. I did #946 to see what others think.

File: examples/psbt.rs
Comment (tcharding) [reply examples/psbt.rs:674]: Legend, cheers.

**Comment by tcharding** [comment]
All green! I rekon this is pretty close to ready-for-review. I still rekon we should be asserting against PSBTs not against hex, do you have an opinion on that?

File: examples/psbt.rs
Comment (DanGould) [reply examples/psbt.rs]: this update only implies the updater knows the particular pubkey and fingerprint to which the xpub is preimage. I think it's safe to assume input output providers have this information. Still, another reviewer ack will be appreciated

**Comment by DanGould** [comment]
We're asserting against PSBTs now 👍

Sincere thanks for all the guidance. I'm thrilled to be able to contribute :D

**Comment by tcharding** [comment]
> We're asserting against PSBTs now

There were a few missed, I pushed a wip patch, you can squash if you like it.

I also pushed a commit that moves the example to `tests/` as is customary for Rust integration tests, run it with `cargo test bip174_psbt_workflow`.

**Comment by tcharding** [comment]
> I'm thrilled to be able to contribute

I'm glad you are enjoying it, this is a mad project with good people working on it. I'm learning heaps hacking on this repo.

**Comment by tcharding** [comment]
Oh and incase you want to know why its in `tests/`: https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html

File: tests/psbt.rs
Comment (sanket1729) [inline tests/psbt.rs:25]: I think we should ask the BDK people to allow re-license the code here to be compatible(If they are different) with what rust-bitcoin currently has. 

I think we can remove the license here(while retaining the credits) if @afilini allows it.

File: tests/psbt.rs
Comment (sanket1729) [inline tests/psbt.rs:47]: I think we should. I use this shorthand for `rust-miniscript`

File: tests/psbt.rs
Comment (sanket1729) [inline tests/psbt.rs:53]: If others have no issues in exporting macros, I think we should export these. I find them really handy while writing tests downstream

File: tests/psbt.rs
Comment (sanket1729) [inline tests/psbt.rs:80]: Yes, this is correct. Can we rename this to `parent_fingerprint` to be more clear?

File: tests/psbt.rs
Comment (sanket1729) [inline tests/psbt.rs:137]: This is confusing as it implies that combining is the last step. If we are trying to test something else, we should create a separate test.

File: tests/psbt.rs
Comment (sanket1729) [inline tests/psbt.rs:443]: Typically tests don't have errors in them. This code is handy, we should use it in the main repo. This would need to be updated for taproot tough.

**Review by sanket1729 — COMMENTED** [review]
Did light and partial review for the first half of the file. Will do a detailed one later

File: tests/psbt.rs
Comment (tcharding) [reply tests/psbt.rs:25]: We have three open PRs that all use this code (#940, #946, and this one). Lets defer the licensing stuff to #946 and apply whatever decision is made there to the other two. I don't think we should merge any of them until this is resolved.

File: tests/psbt.rs
Comment (tcharding) [reply tests/psbt.rs:47]: Done in #951, thanks.

File: tests/psbt.rs
Comment (tcharding) [reply tests/psbt.rs:53]: This is a bit thorny, the `hex_psbt` in `psbt/macros.rs` uses private prelude module so cannot be exported. I couldn't immediately come up with a good clean solution, will put more thought into it later.

File: tests/psbt.rs
Comment (tcharding) [reply tests/psbt.rs:80]: Good idea, leaving this for @DanGould

File: tests/psbt.rs
Comment (tcharding) [reply tests/psbt.rs:137]: Fair point, shall we remove this then @DanGould?

File: tests/psbt.rs
Comment (tcharding) [reply tests/psbt.rs:443]: Agree, this error exists because of the as-yet-unresolved desire to upstream the bdk code. I'm starting to feel like this PR and #940 should be draft until #946 resolves?

File: tests/psbt.rs
Comment (afilini) [reply tests/psbt.rs:25]: See my answer here: https://github.com/bitcoindevkit/bdk/issues/586#issuecomment-1102266447

File: examples/psbt.rs
Comment (apoelstra) [reply examples/psbt.rs]: I think it just didn't occur to me. I have never needed to push a zero to the stack, only FALSE :shrug: 

I'd be happy to accept a new alias.

File: examples/psbt.rs
Comment (tcharding) [reply examples/psbt.rs]: @sanket1729 said this was true somewhere else.

File: tests/psbt.rs
Comment (sanket1729) [reply tests/psbt.rs:80]: To elaborate, the difference between creator/and updater is very fuzzy. In most cases, they are the same entity.

File: tests/psbt.rs
Comment (DanGould) [reply tests/psbt.rs:137]: Shall we? The test is the bip174 defined workflow, of which the lexicographical psbt combiner is a part

Comment (tcharding) [reply tests/psbt.rs:133]: Perhaps we could add a comment explicitly stating that sorting is not part of a typical workflow but is done here for testing purposes?

**Comment by tcharding** [comment]
Hey @DanGould, we can bring these back to life now that MSRV has been bumped and edition 2018 has been enabled. I've got a docs PR to go in first but https://github.com/rust-bitcoin/rust-bitcoin/pull/957 is ready. You could rebase this on top of that and use the `psbt.sign` method if you want to.

File: tests/psbt.rs
Comment (DanGould) [reply tests/psbt.rs:25]: `sign_psbt` code remains a draft because of taproot. We _could_ move forward as-is, including the BDK license & attribution in this file, based on the discussion in https://github.com/bitcoindevkit/bdk/issues/586#issuecomment-1102266447

File: tests/psbt.rs
Comment (DanGould) [reply tests/psbt.rs:439]: #946 closed replaced by #957

File: tests/psbt.rs
Comment (DanGould) [reply tests/psbt.rs]: https://github.com/rust-bitcoin/rust-bitcoin/blob/e557e896dac9579ffc38da3656dbc85ac4820683/tests/psbt.rs#L132-L139

File: tests/psbt.rs
Comment (tcharding) [reply tests/psbt.rs:25]: If you want to push this in before #957 you could pull the ecdsa signing code out of 957 in place of what we did here, and remove the attribution stuff. Saves reviewers bothering with license issues. 

Although, in my opinion, it would be best to rebase this on top of 957 and leave it as draft until 957 goes in but explicitly stating in the PR description that this PR validates/tests the signing code in #957. We can then do the same with #940 and #999. Also this would mean that you, myself, and @dunxen all have a vested interest in getting 957 correct and merged :)

FTR I have previously rebased this work on top of 957 and it all works.

File: tests/psbt.rs
Comment (tcharding) [reply tests/psbt.rs:442]: > This would need to be updated for taproot tough.

After spending today on taproot signing, I got a giggle out of this typo :)

**Comment by DanGould** [comment]
@tcharding rebased on #957 using `psbt.sign`

File: tests/psbt.rs
Comment (tcharding) [inline tests/psbt.rs:1]: Needs a module doc here i.e., `//! PSBT test vectors from BIP-174` or something like that.

File: tests/psbt.rs
Comment (tcharding) [inline tests/psbt.rs:8]: Would be better as:
```
use bitcoin::opcodes::all::OP_PUSHBYTES_0 as OP_0;
```

File: tests/psbt.rs
Comment (tcharding) [inline tests/psbt.rs:9]: We don't put the braces for single imports, one day (TM) we will have tooling to remove stuff like this :)

File: tests/psbt.rs
Comment (tcharding) [inline tests/psbt.rs:416]: We got the greenlight on this already: https://github.com/rust-bitcoin/rust-bitcoin/pull/935#discussion_r854539416

Before merging we should either put an initial patch before this patch that adds the new alias or raise an issue describing what needs to be done (my preference would be the first option). Either way we should remove all the TODOs from this file before attempting to merge (I just wrote them in there for us during development).

File: tests/psbt.rs
Comment (tcharding) [inline tests/psbt.rs:412]: We bumped MSRV already so we can now write this more naturally as:
```
    let sigs: Vec<_> = psbt.inputs[0].partial_sigs.values().collect();
    let script_sig = script::Builder::new()
        .push_opcode(OP_0) // OP_CHECKMULTISIG bug pops +1 value when evaluating so push OP_0.
        .push_slice(&sigs[0].serialize())
        .push_slice(&sigs[1].serialize())
        .push_slice(&psbt.inputs[0].redeem_script.clone().unwrap().serialize())
        .into_script();
```

**Review by tcharding — COMMENTED** [review]
This is some sexy testing, if I do say so myself. I left a couple of minor improvement suggestions. Also, you'll likely want to re-write the PR description, put this into draft, and mention that it's draft because it includes another PR as patch 1. I'll update #957 to mention that this PR build on top of it.

Good work bro

File: tests/psbt.rs
Comment (sanket1729) [reply tests/psbt.rs]: nit: Can you please remove the `Step 9` part from this?

Comment (tcharding) [reply tests/psbt.rs]: > Shall we? The test is the bip174 defined workflow, of which the lexicographical psbt combiner is a part

Do you disagree with  this statement @sanket1729? Or is it that the newly added comment does not fully explain why the `combine_lexicagraphically` call is there?

File: tests/psbt.rs
Comment (tcharding) [reply tests/psbt.rs:20]: Solved it, we only have `Vec` when "std" feature is enabled so we can feature gate on "std" and use `Vec` directly. https://github.com/rust-bitcoin/rust-bitcoin/compare/master...tcharding:06-23-psbt-hex-macro?expand=1

**Comment by DanGould** [comment]
I fixed all the dangling issues I found with this PR. How did you plan to merge this before #957? This PR includes that signing code.

**Comment by tcharding** [comment]
I'll review for you in the morning with fresh eyes.

**Comment by tcharding** [comment]
Sorry for the delay, I forgot about this and went on holidays. I pushed a commit that uses the signing code verbatim from #940 so you don't have to do too much more work thanks to me. Just squash it if you want to go this direction or remove it if you disagree. Will need a rebase to pick up a clippy bug fix, then IMO you can force push hit 'Ready for review' :)

Cheers man

**Comment by DanGould** [comment]
Now this uses the same signer as #940, rebased on master, satisfies the clippy gods.

I pair programmed the latest with @0xBEEFCAF3 and added him as a co-author.

File: tests/psbt.rs
Comment (tcharding) [inline tests/psbt.rs:376]: This has trailing whitespace, the githook is not catching it, I can't work out why.

File: tests/psbt.rs
Comment (tcharding) [inline tests/psbt.rs:384]: Trailing whitespace here also.

File: tests/psbt.rs
Comment (tcharding) [inline tests/psbt.rs:407]: Perhaps we should comment here that finalizing can better be done with `miniscript`?

Comment (tcharding) [reply tests/psbt.rs:407]: If we do we should add the same comment in `examples/ecdsa-psbt.rs`.

File: tests/psbt.rs
Comment (tcharding) [inline tests/psbt.rs:402]: I applied the patches from this PR to the [sign_psbt](https://github.com/rust-bitcoin/rust-bitcoin/pull/957) and was able to implement `GetKey for &[PriavateKey]` based on this code. If/when both this PR and #957 merges this function can become
```rust
/// Signs `psbt` with `keys` if required.
fn sign(mut psbt: Psbt, keys: &[PrivateKey]) -> Psbt {
    let secp = SECP256K1;
    psbt.sign(&keys, &secp);
    psbt
}
```
which smells like success :)

**Review by tcharding — COMMENTED** [review]
This looks good, a few trivial whitespace issues to fix. 

With the whitespace fixes: ACK 210f5ecb

**Comment by tcharding** [comment]
Thanks for sticking at this one @DanGould!

**Comment by tcharding** [comment]
Hey @DanGould, now that the `ecdsa-psbt` module merged and the PSBT signing API improvements, this one should be able to merge. Have you time to check it over and rebase?

**Comment by tcharding** [comment]
I re-named the PR to remove the word "example" because this adds tests not example.

File: tests/psbt.rs
Comment (DanGould) [reply tests/psbt.rs]: This took a little massaging since `&keys` needs to be `GetKeys`. I think it works now

File: bitcoin/tests/psbt.rs
Comment (tcharding) [inline bitcoin/tests/psbt.rs:167]:
```suggestion
                sequence: Sequence::MAX, // Disable nSequence.
```

File: bitcoin/tests/psbt.rs
Comment (tcharding) [inline bitcoin/tests/psbt.rs:159]: nit: With the addition of `use bitcoin::locktime::absolute;

```suggestion
        lock_time: absolute::PackedLockTime::ZERO,
```

File: bitcoin/tests/psbt.rs
Comment (tcharding) [inline bitcoin/tests/psbt.rs:14]: nit: We are in the process of flattening `util`, would be better if the import path did not use `util` (both `psbt` and `bip32` are in scope at the crate root).

**Review by tcharding — APPROVED** [review]
I've said it before but this test is hot. All my suggestions are just nits though I'd like to see `util` removed from the import paths if that's ok.

ACK 421fe924c054ecc4cfd875d0842b0f32f1824324

**Comment by tcharding** [comment]
Will need to be rebased once #1345 goes in.

**Review by tcharding — DISMISSED** [review]
formatted, awesome. Thanks man.

ACK 5459c4f8bd51ebf7059bab6b54324c92bfd11bf0

(Rebase after we merge fix to master still to come.)

**Comment by DanGould** [comment]
cherry-picked https://github.com/rust-bitcoin/rust-bitcoin/pull/1347 so we have a clean history and pass CI before @tcharding wakes up

**Comment by tcharding** [comment]
ACK b8bd31d5a85fb21ad80e1207997ac5e725f9c879

Thanks man!

**Review by tcharding — APPROVED** [review]

**Review by sanket1729 — APPROVED** [review]
ACK b8bd31d5a85fb21ad80e1207997ac5e725f9c879. Thanks for sticking with this PR. This looks awesome after #957

**Pull request merged**
